design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,314 @@
+done by calling the methods on their parent object. Examples:
+
+```python
+import lakefs
+
+repo = lakefs.repository('example')
+branch = repo.branch('main')
+for item in branch.list_objects(prefix='foo/'):
+    if item.path.endswith('.parquet'):
+        print(item.path)
+
+data: bytes = branch.download('datasets/foo/1.parquet')
+branch.upload(data, 'foo/bar/baz.parquet')
+
+# this will work:
+data: bytes = lakefs.repository('example').commit('abc123').download('a/b.txt')
+
+# since commits are immutable, this method will not exist:
+lakefs.repository('example').commit('abc123').upload(data, 'a/b.txt')
+```
+
+## Partial interface definition
+
+### Authentication
+
+```python
+class Client:
+    """
+    Wrapper around lakefs_sdk's client object
+    Takes care of instantiating it from the environment
+    """
+    def __init__(self, **kwargs):
+        self._client = _infer_auth_chain(**kwargs)
+
+# global default client
+DefaultClient: Optional[Client] = None
+
+try:
+    DefaultClient = Client()
+except NoAuthenticationFound:
+    # must call init() explicitly
+    DefaultClient = None
+
+
+def init(**kwargs):
+    global DefaultClient
+    DefaultClient = Client(**kwargs)
+
+```
+
+### model driven interface
+
+
+
+```python
+class Repository:
+    def __init__(self, repository_id: str, client: Client = DefaultClient): ...
+    def get_or_create(self, storage_namespace: str, default_branch_id: str = 'main', include_samples: bool = False) -> Repository: ...
+    def branch(self, branch_id: str) -> Branch: ...
+    def list_branches(self, max_amount: Optional[int], after: str ='', prefix: str ='') -> Generator[Branch]: ...
+    def commit(self, commit_id: str) -> Reference: ...
+    def tag(self, tag_id: str) -> Tag: ...
+    def list_tags(self, max_amount: Optional[int], after: str ='', prefix: str ='') -> Generator[Tag]: ...
+    def metadata(self) -> dict[str, str]: ...
+
+
+class ReadIOBase:
+    def download(self, path: str) -> bytes: ...
+    def list_objects(self, max_amount: Optional[int], after: str ='', prefix: str ='', delimiter: str = '/') -> Generator[ObjectInfo | CommonPrefix]: ...
+    def stat_object(self, path: str) -> ObjectInfo: ...
+    def open(self, path, mode: Literal['r', 'rb'], pre_signed: Optional[bool] = None) -> TextIO | BinaryIO: ...
+
+
+class Reference(ReadIOBase):
+    def __init__(self, repository_id: str, reference_id: str, client: Client = DefaultClient): ...
+    def log(self, max_amount: Optional[int]) -> Generator[Reference]: ...
+    def metadata(self) -> dict[str, str]: ...
+    def commit_message(self) -> str: ...
+    def diff(self, other_ref: str | Reference, max_amount: Optional[int], after: str ='', prefix: str ='', delimiter: str = '/') -> Generator[Change]: ...
+    def merge_into(self, destination_branch_id: str | Branch): ...
+
+
+class WriteIOBase:
+    def upload(self, data: bytes | str | TextIO | BinaryIO, path: str) -> ObjectInfo: ...
```

Comment:
Should _also_ add options to presign and even direct-access uploads.

Reply:
Agree!

Reply:
Added